reference/src/abort-and-assert/clever-errors.md

```diff
@@ -0,0 +1,230 @@
+# Clever Errors
+
+Clever errors are a feature that allows for more informative error messages when an assertion fails
+or an abort is raised. They are a source feature and compile to a `u64` abort code value that
+contains the information needed to access the line number, constant name, and constant value given
+the clever error code and the module that the clever error constant was declared in. Because of
+this, post-processing is required to go from the `u64` abort code value to a human-readable error
+message. This post-processing is automatically performed by the Sui GraphQL server, as well as the
+Sui CLI. If you want to manually decode a clever abort code, you can use the process outlined in
+[Inflating Clever Abort Codes](#inflating-clever-abort-codes) to do so.
+
+> Clever errors include source line information amongst other data. Because of this their value may
+> change due to any changes in the source file (e.g., due to auto-formatting, adding a new module
+> member, or adding a newline).
+
+## Clever Abort Codes
+
+Clever abort codes allow you to use non-u64 constants as abort codes as long as the constants are
+annotated with the `#[error]` attribute. They can be used both in assertions, and as codes to
+`abort`.
+
+```move
+module 0x42::a_module;
+
+#[error]
+const EIsThree: vector<u8> = b"The value is three";
+
+// Will abort with `EIsThree` if `x` is 3
+public fun double_except_three(x: u64): u64 {
+    assert!(x != 3, EIsThree);
+    x * x
+}
+
+// Will always abort with `EIsThree`
+public fun clever_abort() {
+    abort EIsThree
+}
+```
+
+In this example, the `EIsThree` constant is a `vector<u8>`, which is not a `u64`. However, the
+`#[error]` attribute allows the constant to be used as an abort code, and will at runtime produce a
+`u64` abort code value that holds:
+
+1. A set tag-bit that indicates that the abort code is a clever abort code.
+2. The line number of where the abort occured in the source file (e.g., 7).
+3. The index in the module's identifier table for the constant's name (e.g., `EIsThree`).
+4. The index of the constant's value in the module's constant table (e.g., `b"The value is three"`).
+
+In hex, if `double_except_three(3)` is called, it will abort with a `u64` abort code as follows:
+
+```
+0x8000_7000_1000_0000
```

Comment:
```suggestion
0x8000_0007_0001_0000
```
Is this correct? I was surprised that the value was set into the higher bits there.

Reply:
Good catch! I indeed got the order swapped here.